CAM: Mirror Dressup

tarman3 opened the pull request:
Allows to create mirrored path without modeling
User can select axis for mirroring or shape as reference

<img width="1418" height="673" alt="Screenshot_20260221_073123_lossy" src="https://github.com/user-attachments/assets/2a196e2d-f274-4c93-a4af-4607c99b0a0c" />

https://github.com/user-attachments/assets/7d9fcd84-9d95-42aa-8f34-7038b1ca2247

---

## Use cases

### Cut mode not important
Apply _MirrorDressup_
Enable option `KeepBasePath` if needed

<img width="1191" height="315" alt="Screenshot_20260220_093222_hor" src="https://github.com/user-attachments/assets/4db278e7-d8a7-4a20-ae8d-9b10c81fd7dc" />

### Cut mode important
Create copy of operation by `OpCopy` tool or macro [CAMDeepCopy](https://forum.freecad.org/viewtopic.php?t=97562)
Change cut direction in copy and apply _MirrorDressup_

<img width="1224" height="255" alt="1_hor" src="https://github.com/user-attachments/assets/f89838fb-f070-4b19-93af-00dad63ee4db" />

tarman3 committed:
CAM: Mirror Dressup

jffmichi commented:
I would suggest that this Dressup should work similar to the part/draft mirror operation. I.e. the user should be able to choose a plane (instead of x/y) and a base point. Also the mirror Dressup should maybe only output the mirrored path, not the original (if the user also wants the original it can be manually added to the operations again,  or maybe a checkbox to select if outputting both or only mirrored)? I don't like having the same mirror things behave totally different.

Without looking at the code, did you invert the cutting direction to keep CW/CCW direction the same?

tarman3 commented:
> Is this fixing a known issue?

No
Yesterday I get task to mill two pieces. One of them mirrored
I suggested that it would be a useful function and interesting not only to me

> Without looking at the code, did you invert the cutting direction to keep CW/CCW direction the same?

No
To get correct direction need switch direction in base operation

> I.e. the user should be able to choose a plane (instead of x/y) and a base point (instead of separate x/y properties)

No. I have another point of view )
I dont want to create some extra plain or point
Maybe will add some relation to stock or model, e.g. something like `UseMiddleOfStock`

Is my first implementation
A lot of bugs inside
But at this time I have only one main question
**Is this feature interesting for others or no?**

jffmichi commented:
Have a look at the mirroring tool in the Part workbench. You don't have to create a plane if you want to use one of the axis aligned ones but you can select one if you need to:

![grafik](https://github.com/user-attachments/assets/116bd7fa-5669-4d74-b17f-70f9419a3097)
![grafik](https://github.com/user-attachments/assets/61b2f826-96fb-42f5-acd0-6c4e824e8ccb)

sliptonic commented:
> > Is this fixing a known issue?
> 
> No Yesterday I get task to mill two pieces. One of them mirrored I suggested that it would be a useful function and interesting not only to me

it may be a useful feature but needs more discussion and input before implementation.
FreeCAD has mirror functions in Draft, Part, BIM, and Part Design already.  It's a mess because they all have very different UIs and workflows which confuse users moving between workbenches.  We don't want to implement yet-another mirror workflow, especially if it isn't fixing an issue that has some consensus.  If there  isn't broad consensus about usefulness, poorly implemented features just sit forever and nobody ever improves them.

Please don't misunderstand me.  I'm NOT saying this is a bad idea.  It might be a very good idea.  I'm saying don't rush to implement the feature without getting some input from the community.  This PR is a very good prototype to show people what you're suggesting.  

I advise:

1) open an issue and clearly describe the problem you're trying to solve. Mention this draft PR so it's linked.
2) ping DWG for opinions on UI design and workflow.
3) ping CWG for input about similar features in other CAM applications.

tarman3 commented:
@sliptonic 
1, 2, 3 ))
sorry, but probably better to close this PR right know )))

I understand that you are trying to bring Freecad CAM to perfection, but I'm definitely not an assistant here
I think this is a task for professional developers

sliptonic commented:
> @sliptonic 1, 2, 3 )) sorry, but probably better to close this PR right know )))
> 
> I understand that you are trying to bring Freecad CAM to perfection, but I'm definitely not an assistant here I think this is a task for professional developers
> 
> I understand that UI/UX is very important, but this is an area for professionals Discussion of each little button by group DWG generally beyond my understanding I need to solve simple task, here, now, and maybe tomorrow

There are are people here who develop software professionally but most are just like you.  They are contributing to FreeCAD as volunteers. You don't need to be an expert on UI, that's what we have a DWG. They are good people who can offer practical suggestions.

The [contributing process](https://github.com/FreeCAD/FreeCAD/blob/main/CONTRIBUTING.md#6-process) is designed to separate the discussion about a problem (issue) from the solution implemented to fix it (PR).  Those are two different discussions.  As maintainers, we're not jerks about enforcing it.  But there's good reasons for having two separate discussions.

When well-meaning contributors jump directly to the implementation, it has two big drawbacks.   First, the developer can invest a lot of time on something that gets rejected. That sucks.

Second, it puts the maintainers in a very hard position.  They have to choose between offending someone who's just trying to help by rejecting a working but poorly implemented solution or accepting  it and burdening other overworked contributors with cleaning up the mess. 

I'm not trying to bring CAM to perfection.  I'm trying to have a development process that maximizes contributions and minimizes pain.

 You're doing really great stuff.  Just don't be a lone wolf.  Be part of the team.

sliptonic commented:
I won't close it but let's leave it as draft.

tarman3 commented:
Fixed problem with automatic recompute
But find new

Can not remove dressup after create
Need save, close and open document. Only after that can remove dressup

tarman3 commented:
Fixed
Recompute and delete working

tarman3 commented:
Maybe someone know the elegant way to get correct Path direction

GeneGH commented:
I have not decided how useful this might be, but it seems much more dangerous than most things in FreeCAD. A mistake by either the program or the user can lead to disaster in the shop.

This is true for everything in CAM, of course, but adding a potentially large extra zone for processing (outside the basic model) seems to be particularly suited for causing havoc.

It may be good to add some enhanced "sanity checking".

tarman3 commented:
Probably you work with machine which have a little work area

For me program like this is normal task, so I am not share your fears
![Screenshot_20250607_163511_lossy](https://github.com/user-attachments/assets/7702ca4b-369a-4276-9e18-d890e5c38840)

tarman3 commented:
Instead of use Base Operation directly, now create `Deep Recursive Copy` with set expressions for almost all properties

Now we can create mirror of complicated Operations and change direction of the Path

![Screenshot_20250608_133930_lossy](https://github.com/user-attachments/assets/3aac585e-eb14-4fd0-a134-f4818c06c940)

https://github.com/user-attachments/assets/eac854a0-df49-4e1c-92d2-542ac347a9e2

tarman3 commented:
Fixed for `Profile` with empty `Base`

sliptonic commented:
@tarman3 Can you check and address the issues flagged by the github-advanced-security  bot.

tarman3 commented:
I saw that and have a lot of this messages in my other PR
but I do not understand what is this mean

` Mixing implicit and explicit returns may indicate an error as implicit returns always return None`

sliptonic commented:
Roadmap has discussion of need for 'array-like' toolpaths.  
Moving this to draft until we agree on direction

tarman3 commented:
Fixes recursive searching of base operation
Now it should process any object which contain `Path`

tarman3 commented:
Different cases requires different approaches
Removed feature to automatically creates recursive copy of operation from _MirrorDressup_
Changed description in top post

Connor9220 commented:
@Connor9220 review

tarman3 commented:
Properties `MirrorOffsetX` and `MirrorOffsetY` replaced by PropertyVectorDistance `Offset`

<img width="1263" height="523" alt="Screenshot_20260220_091644_lossy" src="https://github.com/user-attachments/assets/715f8161-ae76-43e1-9d10-7ebe8e70f5e4" />
